internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -155,38 +179,156 @@ func (f *fleetGateway) Run(ctx context.Context) error {
+				if len(actions) > 0 {
+					f.actionCh <- actions
+				}
+			}()
+
+		}
+	}
+}
 
-			// Execute the checkin call and for any errors returned by the fleet-server API
-			// the function will retry to communicate with fleet-server with an exponential delay and some
-			// jitter to help better distribute the load from a fleet of agents.
-			resp, err := f.doExecute(ctx, backoff)
-			if err != nil {
+func (f *fleetGateway) performCancellableCheckin(ctx context.Context, initialState state.State, stateUpdates <-chan state.State, backoff backoff.Backoff, dbouncer debouncer) checkinResult {
+	checkinID, _ := uuid.NewV4()
+	checkinCtx, cancelCheckinCtxFunc := context.WithCancel(ctx)
+	checkinCtx = context.WithValue(checkinCtx, CheckinIDKey, checkinID.String())
+	f.log.Debugf("starting checkin %q", checkinID.String())
+	defer func() {
+		// make sure we don't leak contexts whatever happens
+		cancelCheckinCtxFunc()
+	}()
+
+	// Execute the checkin call asynchronously. For any errors returned by the fleet-server API
+	// the function will retry to communicate with fleet-server with an exponential delay and some
+	// jitter to help better distribute the load from a fleet of agents.
+	resCheckinChan := f.doExecuteAsync(checkinCtx, backoff, initialState)
+
+	// this variable will hold the latest state we got from coordinator
+	var updatedState *state.State
+	var debounceElapsed bool
+
+	for {
+		select {
+		case <-ctx.Done():
+			return checkinResult{err: ctx.Err()}
+		case checkinResult := <-resCheckinChan:
+			return checkinResult
+		case <-dbouncer.Elapsed():
+			debounceElapsed = true
+			if updatedState != nil && !reflect.DeepEqual(updatedState, &initialState) {
+				// we detected that the updated State is different from the one of the current checkin, cancel and return the needNewCheckinError
+				f.log.Debugf("Detected an updated state at the end of the debounce, cancelling ongoing checkin %q", checkinID)
+				cancelCheckinTimeoutCtx, cancelCancelCheckin := context.WithTimeout(ctx, f.cancelTimeout)
+				defer cancelCancelCheckin()
+				f.cancelCheckin(cancelCheckinTimeoutCtx, checkinID.String(), cancelCheckinCtxFunc, resCheckinChan)
+				return checkinResult{err: &needNewCheckinError{newState: *updatedState}}
+			}
+			f.log.Debug("No new state detected at the end of the debounce")
+		case newState := <-stateUpdates:
+			if !debounceElapsed {
+				f.log.Debugf(
+					"Updated state (Agent state: %q) received but current checkin %s not cancelled because we are still within debounce",
+					newState.State,
+					checkinID,
+				)
+				updatedState = &newState
 				continue
 			}
 
-			actions := make([]fleetapi.Action, len(resp.Actions))
-			copy(actions, resp.Actions)
-			if len(actions) > 0 {
-				f.actionCh <- actions
-			}
+			f.log.Infof(
+				"Received updated state (Agent state: %q) when checkin is ongoing past configured debounce. Cancelling previous checkin %q and starting a new one.",
+				newState.State,
+				checkinID.String(),
+			)
```

Comment:
I wasn't sure if we wanted to see when we are cancelling a checkin with a given state because of an update.
If we are not interested this can be moved to `Debugf`